README.txt
        
                             C A N T E R A

                              release 1.7

                               11/2006

      Copyright (c) 2001-2006 California Institute of Technology

***************************************************************
***                                                         
***   For information on installing and using Cantera, see  
***        http://www.cantera.org/Documentation             
***                                                         
***************************************************************

License information
===================

See the file "License.txt" for information on the terms & conditions
for usage, and a DISCLAIMER OF ALL WARRANTIES.

All trademarks referenced herein are property of their respective
holders.


Web sites
==========

The main Cantera web site is http://www.cantera.org. This serves as a
gateway to three other web sites:

1. The Cantera User's Group. http://groups.yahoo.com/groups/cantera.   
   This site has a message board, and some miscellaneous files and utilities. 

2. The Cantera Sourceforge site. Distribution of the Cantera source code is
   done using Sourceforge. The site is http://sourceforge.net/projects/cantera.

3. The Cantera wiki. http://www.cantera.org/wiki.



ChangeLog
2006-11-14 00:40  dggoodwin

	* configure, configure.in, Cantera/src/MultiPhaseEquil.cpp: minor
	  cleanup

2006-11-13 11:55  dggoodwin

	* Cantera/src/CVodesIntegrator.h, ext/f2c_libs/arith.h,
	  win32/vc7/Sundials/CVODES/CVODES.vcproj,
	  win32/vc7/Sundials/NVEC_SER/NVEC_SER.vcproj,
	  win32/vc7/Sundials/SUNDIALS_SHARED/SUNDIALS_SHARED.vcproj,
	  win32/vc7/cantera/cantera.vcproj: minor cleanup

2006-11-13 11:46  dggoodwin

	* win32/vc7/Sundials/sundials/sundials_config.h: added
	  sundials_config.h so that users can build sundials without having
	  to run configure

2006-11-13 08:39  dggoodwin

	* ChangeLog, preconfig, ext/f2c_libs/arith.h,
	  test_problems/python/runtest: minor changes

2006-11-12 20:23  dggoodwin

	* License.rtf, preconfig, winconfig.h, ext/f2c_libs/arith.h: minor
	  cleanup

2006-11-12 20:22  dggoodwin

	* win32/vc7/demos/: NASA_coeffs.vcproj, combustor.vcproj,
	  flamespeed.vcproj, kinetics1.vcproj: added example project files

2006-11-12 20:21  dggoodwin

	* win32/vc7/: SetupCantera/SetupCantera.vdproj,
	  SetupCantera/SetupCanteraLite.vdproj, cantera.sln,
	  Sundials/CVODES/CVODES.vcproj,
	  Sundials/CVODES_DENSE/CVODES_DENSE.vcproj,
	  Sundials/IDA/IDA.vcproj, Sundials/KINSOL/KINSOL.vcproj,
	  Sundials/NVEC_SER/NVEC_SER.vcproj,
	  Sundials/SUNDIALS_SHARED/SUNDIALS_SHARED.vcproj,
	  cantera/cantera.vcproj, demos/CanteraDemos.sln,
	  demos/Rankine.vcproj, demos/demo.vcproj, zeroD/zeroD.vcproj:
	  update of project files

2006-11-11 06:12  dggoodwin

	* win32/vc7/: cantera.sln, cantera_no_sundials.sln,
	  SetupCantera/SetupCantera.vdproj, cantera/cantera.vcproj: added
	  no_sundials version

2006-11-11 06:10  dggoodwin

	* win32/vc7/Sundials/: Sundials.sln,
	  CVODES_DENSE/CVODES_DENSE.vcproj: added cvodes_dense

2006-11-11 05:23  dggoodwin

	* preconfig, Cantera/src/Makefile.in,
	  Cantera/src/zeroD/Makefile.in: minor changes to how dependencies
	  are handled in Makefiles

2006-11-10 17:38  dggoodwin

	* configure, configure.in: fixed null username

2006-11-10 17:37  dggoodwin

	* tools/src/finish_install.py.in: added writing PYTHONPATH to
	  setup_Cantera for min_Python

2006-11-10 05:48  dggoodwin

	* Cantera/src/CVodesIntegrator.h: minor cleanup

2006-11-09 05:04  dggoodwin

	* Cantera/python/examples/fuel_cells/: sofc.cti, sofc.py: added
	  fuel cell demo

2006-11-08 16:19  dggoodwin

	* Cantera/src/ThermoPhase.h: minor cleanup

2006-11-08 15:58  dggoodwin

	* Makefile.in, preconfig, Cantera/python/Cantera/OneD/onedim.py,
	  ext/f2c_libs/arith.h: install Makefiles for python demos

2006-11-08 15:14  dggoodwin

	* Cantera/python/examples/: Makefile, equilibrium/Makefile,
	  equilibrium/stoich.py, flames/Makefile, gasdynamics/Makefile,
	  kinetics/Makefile, liquid_vapor/Makefile, misc/Makefile,
	  reactors/Makefile, surface_chemistry/Makefile,
	  transport/Makefile: added Makefiles to run all demos

2006-11-07 13:03  dggoodwin

	* Cantera/src/zeroD/ReactorNet.h: added verbose, setVerbose

2006-11-07 13:02  dggoodwin

	* Cantera/src/zeroD/ReactorBase.h: added setName

2006-11-07 05:47  dggoodwin

	* Cantera/src/: State.h, State.cpp: added method concentration(int
	  k)

2006-11-06 07:27  dggoodwin

	* Cantera/cxx/demos/NASA_coeffs.cpp: added NASA polynomial example

2006-11-06 07:20  dggoodwin

	* Cantera/src/: Constituents.h, NasaPoly1.h, NasaThermo.h,
	  SpeciesThermo.h, SpeciesThermoFactory.cpp,
	  SpeciesThermoInterpType.h: added support for modifying NASA
	  polynomial coefficients

2006-11-05 07:54  dggoodwin

	* data/inputs/methane_pox_on_pt.cti: added mechanism for partial
	  oxidation of methane on platinum from the Deutschmann group

2006-11-03 12:45  dggoodwin

	* winconfig.h: updated

2006-11-03 12:44  dggoodwin

	* win32/vc7/Sundials/: CVODES/CVODES.vcproj, IDA/IDA.vcproj,
	  KINSOL/KINSOL.vcproj, NVEC_SER/NVEC_SER.vcproj,
	  SUNDIALS_SHARED/SUNDIALS_SHARED.vcproj: initial import

2006-11-02 17:32  dggoodwin

	* win32/vc7/SetupCantera/SetupCantera.vdproj,
	  win32/vc7/SetupCantera/SetupCanteraLite.vdproj,
	  win32/vc7/ctmatlab/runmlab.cmd, Cantera/python/winsetup.py,
	  License.txt: updated for Cantera 1.7

2006-10-31 11:09  dggoodwin

	* Cantera/src/ReactionStoichMgr.cpp: fixed bug that resulted in
	  incorrect computation of product concentration products when rxn
	  has 4 or more product molecules and one or more has a stoich
	  coeff > 1

2006-10-30 11:30  dggoodwin

	* configure, configure.in, preconfig, ext/f2c_libs/arith.h,
	  tools/src/finish_install.py.in: modified for cygwin

2006-10-27 13:34  dggoodwin

	* configure, configure.in, tools/src/finish_install.py.in: mods for
	  cygwin

2006-10-22 18:13  dggoodwin

	* Cantera/src/CVodesIntegrator.cpp: sundials version detection

2006-10-22 17:58  dggoodwin

	* preconfig: debug off is noe the default

2006-10-22 17:56  dggoodwin

	* Cantera/src/: ChemEquil.cpp, MultiPhaseEquil.cpp:  moved
	  declaration of nCutoff to avoid jumping over it, causing a
	  compile failure with gcc 4.0.1 on linuxChemEquil.cpp

2006-10-22 17:51  dggoodwin

	* tools/templates/cxx/demo.mak.in: removed FLIBS

2006-10-20 14:56  hkmoffa

	* test_problems/min_python/Makefile.in: Added -s to make tests.

2006-10-20 14:55  hkmoffa

	* test_problems/cathermo/HMW_graph_GvI/.cvsignore: Added a file.

2006-10-20 14:54  hkmoffa

	* test_problems/fracCoeff/runtest: Added an exit variable.

2006-10-20 14:47  hkmoffa

	* test_problems/ChemEquil_gri_matrix/output_blessed.txt: Slight
	  change to the blessed solution.

2006-10-20 14:43  hkmoffa

	* configure, configure.in, test_problems/Makefile.in: Added new
	  test directories.

2006-10-20 14:42  hkmoffa

	* Cantera/src/sort.h: Added an Id stamp.

2006-10-20 14:41  hkmoffa

	* Cantera/src/misc.cpp: Fixed an error in the LogGroup system.

2006-10-20 14:40  hkmoffa


ext/tpx/Sub.h
/*
 * This is the base substance class from which all substances are derived
 * 
 * Kate Talmazan: SURF -- July, 1995
 *   original implementation of this class and all derived classes from 
 *   formulas given in TPSI. Implementation of P(Rho, T), cv0(T), ldens(T),
 *   and Psat(T) for all substances in TPSI.f
 *
 * Dave Goodwin: Fall, 1996
 *   functions for u, h, s, f, g; 
 *   functions to set state
 *   error handling
 *   documentation 
 *
 *   Sept., 2001: minor modifications to use with Cantera
 *
 */

#ifndef TPX_SUB_H
#define TPX_SUB_H

#include <iostream>
#include <string>
using namespace std;

namespace tpx {

    class TPX_Error {
    public:
        TPX_Error(string p, string e) {
            ErrorMessage = e;
            ErrorProcedure = p;
        }
        virtual ~TPX_Error(){}
        static string ErrorMessage;
        static string ErrorProcedure;
    };


    const double OneAtm = 1.01325e5;
    const double Liquid = 0.0;
    const double Vapor = 1.0;

    const int TV = 12, HP = 34, SP = 54, PV = 42, TP = 14, UV = 62, 
              ST = 51, SV = 52, UP = 64, VH = 23, TH = 13, SH = 53, 
              PX = 47, TX = 17;

    const int VT = -12, PH = -34, PS = -54, VP = -42, PT = -14, VU = -62,
              TS = -51, VS = -52, PU = -64, HV = -23, HT = -13, HS = -53,
              XP = -47, XT = -17;

    const int NoConverge = -900;
    const int GenError = -901;
    const int InvalidInput = -902;
    const int TempError = -800;
    const int PresError = -801;
    const int CKError = -802;


    const int Pgiven = 0, Tgiven = 1;

    const int EvalH = 3;  const int EvalS = 5;
    const int EvalU = 6;  const int EvalV = 2;  const int EvalP = 4;
    const int EvalT = 1;  const int EvalX = 7;  const int EvalF = 8;
    const int EvalG = 9; const int EvalC = 10;
    const int EvalM = 11; const int EvalN = 12; const int EvalMW = 13;
    const int EvalEA = 14;
    const int EvalCdot = 15; const int EvalDdot = 16; const int EvalWdot = 17;
    const int EvalTchem = 18; const int EvalRgas = 19;

    const double Undef = 999.1234;

    string errorMsg(int flag);

    class Substance { 
    public:
        Substance() { 
            T = Undef;
            Rho = Undef;
            Tslast = Undef; 
            Rhf = Undef; 
            Rhv = Undef; 
            Pst = Undef; 
            Err = 0;
            m_energy_offset = 0.0;
            m_entropy_offset = 0.0;
        }
	virtual ~Substance(){}

        void setStdState(double h0 = 0.0, double s0 = 0.0,
            double t0 = 298.15, double p0 = 1.01325e5) {
            Set(TP, t0, p0);
            double hoff = h0 - h();
            double soff = s0 - s();
            m_entropy_offset = soff;
            m_energy_offset = hoff;
        }

        // information about a substance:

        virtual double MolWt()=0;          // molecular weight, kg/kmol
        virtual double Tcrit()=0;          // critical temperature, K
        virtual double Pcrit()=0;          // critical pressure, Pa
        virtual double Vcrit()=0;          // critical specific vol, m^3/kg
        virtual double Tmin()=0;           // min. temp for which equations valid
        virtual double Tmax()=0;           // max. temp for which equations valid
        virtual char * name() = 0;         // name
        virtual char * formula() = 0;      // chemical formula
        
        // properties:
        
        double P();                        // pressure, Pa
        double Temp() {return T;}          // temperature, K
        double v()                         // specific vol, m^3/kg
            {return prop(EvalV);}
        double u()                         // int. energy, J/kg
            {return prop(EvalU);}
        double h()                         // enthalpy, J/kg
            {return prop(EvalH);}
        double s()                         // entropy, J/kg/K
            {return prop(EvalS);}
        double f()                         // Helmholz function, J/kg
            {return u() - T*s();}
        double g()                         // Gibbs function, J/kg
            {return h() - T*s();}

        virtual double cv() {
            double Tsave = T, dt = 1.e-4*T;
            set_T(Tsave - dt);
            double s1 = s();
            set_T(Tsave + dt);
            double s2 = s();
            set_T(Tsave);
            return T*(s2 - s1)/(2.0*dt);
        }

        virtual double cp() {
            double Tsave = T, dt = 1.e-4*T;
            double p0 = P();
            Set(TP, Tsave - dt, p0);
            double s1 = s();
            Set(TP, Tsave + dt, p0);
            double s2 = s();
            Set(TP, Tsave, p0);
            return T*(s2 - s1)/(2.0*dt);
        }

        virtual double thermalExpansionCoeff() {
            double Tsave = T, dt = 1.e-4*T;
            double p0 = P();
            Set(TP, Tsave - dt, p0);
            double v1 = v();
            Set(TP, Tsave + dt, p0);
            double v2 = v();
            Set(TP, Tsave, p0);
            return (v2 - v1)/((v2 + v1)*dt);
        }

        virtual double isothermalCompressibility() {
            double Psave = P(), dp = 1.e-4*Psave;
            Set(TP, T, Psave - dp);
            double v1 = v();
            Set(TP, T, Psave + dp);
            double v2 = v();
            Set(TP, T, Psave);
            return -(v2 - v1)/((v2 + v1)*dp);
        }


        // saturation properties

        double Ps();
        virtual double dPsdT();          // d(Psat)/dT, Pa/K
        double Tsat(double p);             // saturation temp at p
        double x();                        // vapor mass fraction
        int TwoPhase();                    // =1 if vapor/liquid, 0 otherwise
        virtual double Pp()=0;
        double hp() {return up() + Pp()/Rho;} 
        double gp() {return hp() - T*sp();}
        
        double prop(int ijob);
        void set_TPp(double t0, double p0);    // set T and P

        
        // functions to set or change state:

        void Set(int XY, double x0, double y0);
        void Set_meta(double phase, double pp);	 

        int Error() {
            return Err;
        } 


    protected:   

	double T, Rho;
	double Tslast, Rhf, Rhv;
	double Pst;
	int Err;
        double m_energy_offset;
        double m_entropy_offset;
        string m_name;
        string m_formula;

        //	virtual double Xm(int k) { return 1.0;}
        //virtual int Species() { return 1;}

	virtual double ldens()=0;
	virtual double Psat()=0;           // saturation pressure, Pa
	virtual double up()=0;
	virtual double sp()=0;
	virtual int ideal(){return 0;}     // added 9/2/98; default is false
	double vp() {return 1.0/Rho;}
	int Lever(int itp, double sat, double val, int ifunc);
	void update_sat();

	void set_Err(int ErrFlag) {
            if (!Err) {
                Err = ErrFlag;
                //throw TPX_Error(""errorMsg(Err));
            }
	}
	void clear_Err() {Err = 0;}


private:
	  void set_Rho(double r0);
	  void set_T(double t0);
	  void set_v(double v0);
	  void BracketSlope(double p);
	  double lprop(int ijob);
	  double vprop(int ijob);
	  void set_xy(int if1, int if2, double X, double Y, 
		  double atx, double aty, double rtx, double rty);
    };

    void Error(char *message, int flag, double val=Undef);
    void Mess(char *message);

}


#endif
